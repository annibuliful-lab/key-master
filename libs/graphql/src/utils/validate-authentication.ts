import { prismaClient, redisClient } from '@key-master/db';
import type { IJwtAuthInfo } from '../@types/auth';
import { verify } from './jwt';

const CACHE_EXPIRE = 3600;
interface IValidateAuthenticationParam {
  token: string;
  projectId?: string | null;
  cacheExpire?: number;
}
export const validateAuthentication = async ({
  token,
  projectId = null,
  cacheExpire = CACHE_EXPIRE,
}: IValidateAuthenticationParam): Promise<
  | { userId: string; permissions: string[]; role: string | null }
  | 'FORBIDDEN'
  | null
> => {
  const { isValid, userInfo } = verify<IJwtAuthInfo>(token);

  if (!isValid) {
    return null;
  }

  // get cache user
  if (userInfo.userId && !projectId) {
    const key = `${userInfo.userId}-null`;
    const authInfo = await redisClient.get(key);
    if (authInfo) {
      return JSON.parse(authInfo);
    }
  }

  // get cache user with projectId from redis
  if (userInfo.userId && projectId) {
    const key = `${userInfo.userId}-${projectId}`;
    const authInfo = await redisClient.get(key);

    if (authInfo) {
      return JSON.parse(authInfo);
    }
  }

  let permissions: string[] = [];
  let role: string | null = null;
  const user = await getUserInfo(userInfo.userId);

  if (projectId) {
    const userPermissions = await getUserPermissions({
      projectId,
      userId: user.id,
    });

    // prevent fobidden permission and role in selected project
    if (!userPermissions) {
      return 'FORBIDDEN';
    }

    // prevent fobidden permission and role in selected project
    if (userPermissions.permissions.length === 0) {
      return 'FORBIDDEN';
    }

    permissions = userPermissions.permissions;

    role = userPermissions.role;

    // remove duplicate user id key
    await redisClient.del(`${user.id}-null`);

    // cache user id with project id
    await redisClient.set(
      `${user.id}-${projectId}`,
      JSON.stringify({
        userId: user.id,
        permissions,
        role,
      }),
      'EX',
      cacheExpire
    );
  }

  if (!user) {
    return null;
  }

  const authInfo = {
    userId: user.id,
    permissions,
    role,
  };

  // set cache for user auth
  await redisClient.set(
    `${user.id}-${projectId}`,
    JSON.stringify(authInfo),
    'EX',
    cacheExpire
  );

  return authInfo;
};

const getUserInfo = (id: string) => {
  return prismaClient.user.findUnique({
    select: {
      id: true,
    },
    where: {
      id,
    },
  });
};

interface IGetUserPermissionsParam {
  projectId: string;
  userId: string;
}
const getUserPermissions = async ({
  projectId,
  userId,
}: IGetUserPermissionsParam): Promise<{
  userId: string;
  permissions: string[];
  role: string;
} | null> => {
  const cacheUserPermissions = await redisClient.get(`${userId}-${projectId}`);

  if (cacheUserPermissions) {
    return JSON.parse(cacheUserPermissions);
  }

  const userPermissions = await prismaClient.projectRoleUser.findFirst({
    select: {
      role: {
        select: {
          role: true,
          rolePermissions: {
            select: {
              permission: {
                select: {
                  permission: true,
                },
              },
            },
          },
        },
      },
    },
    where: {
      projectId,
      project: {
        deletedAt: null,
      },
      role: {
        deletedAt: null,
        rolePermissions: {
          some: {
            permission: {
              deletedAt: null,
            },
            deletedAt: null,
          },
        },
      },
      userId: userId,
    },
  });

  if (!userPermissions) {
    return null;
  }

  return {
    userId,
    role: userPermissions.role.role,
    permissions: userPermissions.role.rolePermissions.map(
      (rolePermission) => rolePermission.permission.permission
    ),
  };
};

export const getAllPermissions = async () => {
  return (
    await prismaClient.permission.findMany({
      select: {
        permission: true,
      },
      where: {
        deletedAt: null,
      },
    })
  ).map((p) => p.permission);
};
