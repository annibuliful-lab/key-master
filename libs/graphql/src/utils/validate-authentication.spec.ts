import { getUserPermissions } from './validate-authentication';
it('returns user permissions with project id', async () => {
  expect(
    await getUserPermissions({
      userId: 'TEST_USER_A_ID',
      projectId: 'TEST_PROJECT_ID',
      skipCache: true,
    })
  ).toMatchSnapshot();
});
