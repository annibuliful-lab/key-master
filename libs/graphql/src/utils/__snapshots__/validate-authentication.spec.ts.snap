// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns user permissions with project id 1`] = `
Object {
  "permissions": Array [
    "USER_WRITE",
    "USER_READ",
    "PROJECT_READ",
    "PROJECT_WRITE",
    "ROLE_READ",
    "ROLE_WRITE",
    "PROJECT_ROLE_READ",
    "PROJECT_ROLE_WRITE",
    "PROJECT_USER_READ",
    "PROJECT_USER_WRITE",
    "PERMISSION_READ",
    "PROJECT_DELETE",
    "PROJECT_ROLE_USER_WRITE",
    "PROJECT_ROLE_USER_READ",
    "PROJECT_ORGANIZATION_USER_READ",
    "PROJECT_ORGANIZATION_USER_WRITE",
    "PROJECT_ORGANIZATION_WRITE",
    "PROJECT_ORGANIZATION_READ",
    "PROJECT_ROLE_PERMISSION_WRITE",
    "PROJECT_ROLE_PERMISSION_READ",
    "ORGANIZATION_KEY_MANAGEMENT_WRITE",
    "ORGANIZATION_KEY_MANAGEMENT_READ",
    "KEY_MANAGEMENT_WRITE",
    "KEY_MANAGEMENT_READ",
    "KEY_MANAGEMENT_UPDATE_PIN",
    "USER_ACTIVITY_READ",
  ],
  "role": "OWNER",
  "userId": "TEST_USER_A_ID",
}
`;
